src/uu/comm/src/comm.rs

```diff
@@ -39,23 +40,55 @@ fn column_width(col: &str, opts: &ArgMatches) -> usize {
     }
 }
 
-fn ensure_nl(line: &mut String) {
-    if !line.ends_with('\n') {
-        line.push('\n');
+#[repr(u8)]
+#[derive(Clone, Copy)]
+enum LineEnding {
+    Newline = b'\n',
+    Nul = 0,
+}
+
+impl From<LineEnding> for u8 {
+    fn from(line_ending: LineEnding) -> Self {
+        line_ending as Self
+    }
+}
+
+impl From<bool> for LineEnding {
+    fn from(is_zero_terminated: bool) -> Self {
+        if is_zero_terminated {
+            Self::Nul
+        } else {
+            Self::Newline
+        }
+    }
+}
+
+impl Display for LineEnding {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Newline => writeln!(f),
+            Self::Nul => write!(f, "\0"),
+        }
     }
 }
 
 enum LineReader {
-    Stdin(Stdin),
-    FileIn(BufReader<File>),
+    Stdin(Stdin, LineEnding),
+    FileIn(BufReader<File>, LineEnding),
 }
 
 impl LineReader {
-    fn read_line(&mut self, buf: &mut String) -> io::Result<usize> {
-        match *self {
-            Self::Stdin(ref mut r) => r.read_line(buf),
-            Self::FileIn(ref mut r) => r.read_line(buf),
+    fn read_line(&mut self, buf: &mut Vec<u8>) -> io::Result<usize> {
+        let (result, line_ending) = match *self {
+            Self::Stdin(ref mut r, le) => (r.lock().read_until(le.into(), buf), le.into()),
```

Comment:
This might be easier if the data type is different. For example:
```rust
struct LineReader {
    line_ending: LineEnding,
    input: Input,
}

enum Input {
    Stdin(Stdin),
    FileIn(BufReader<File>),
}
```

Reply:
Good point, this looks cleaner. Thanks. I will change it.